File: frontend/src/app/models/game.models.ts
export type Player = 'X' | 'O';
export type GameMode = 'TwoPlayer' | 'Computer';
export type GameStatus = 'InProgress' | 'Won' | 'Draw';

export interface Move {
  player: Player;
  row: number;
  column: number;
  moveNumber: number;
}

export interface Scoreboard {
  xWins: number;
  oWins: number;
  draws: number;
}

export interface GameState {
  gameId: string;
  board: Array<Player | null>;
  currentPlayer: Player;
  mode: GameMode;
  status: GameStatus;
  winner: Player | null;
  winningCells: number[];
  moveHistory: Move[];
  scoreboard: Scoreboard;
}

export interface ApiError {
  code: string;
  message: string;
}

File: backend/src/TicTacToe.Api/Domain/GameModels.cs
namespace TicTacToe.Api.Domain;

public enum Player
{
    X,
    O
}

public enum GameMode
{
    TwoPlayer,
    Computer
}

public enum GameStatus
{
    InProgress,
    Won,
    Draw
}

public sealed record Move(Player Player, int Row, int Column, int MoveNumber);

public sealed class GameSession
{
    public required Guid Id { get; init; }
    public GameMode Mode { get; set; }
    public Player CurrentPlayer { get; set; } = Player.X;
    public GameStatus Status { get; set; } = GameStatus.InProgress;
    public Player? Winner { get; set; }
    public List<int> WinningCells { get; } = [];
    public Player?[] Board { get; } = new Player?[9];
    public List<Move> Moves { get; } = [];
    public bool ScoreApplied { get; set; }
}

public sealed class Scoreboard
{
    public int XWins { get; set; }
    public int OWins { get; set; }
    public int Draws { get; set; }
}
